src/site/content/en/blog/porting-libusb-to-webusb/index.md

```diff
@@ -0,0 +1,450 @@
+
+There are a couple of problems with this approach on the web.
+
+First of all, WebUSB doesn't and cannot expose raw handles of the underlying devices, so polling those directly is not an option. Second, libusb uses `eventfd` and `pipe` APIs for other events as well as for handling transfers on operating systems without raw device handles, but `eventfd` is not currently supported in Emscripten, and `pipe`, while supported, [currently doesn't conform to the spec](https://github.com/emscripten-core/emscripten/issues/13214) and can't wait for events.
+
+Finally, the biggest problem is that the web has its own event loop. This global event loop is used for any external I/O operations (including `fetch`, timers, or, in this case, WebUSB), and it invokes event or `Promise` handlers whenever corresponding operation has finished. Executing another, nested, infinite event loop will block the browser's event loop from ever progressing, which means that not only will the UI become unresponsive, but also that the code will never get notifications for the very same I/O events it's waiting for. This usually results in a deadlock, and that's what happened when I tried to use libusb in a demo, too - the page froze.
+
+Like with other blocking I/O, to port such event loops to the web, developers need to find a way to run those loops without blocking the main thread. One way is to refactor the application to handle I/O events in a separate thread and pass the results back to the main one. The other is to use Asyncify to pause the loop and wait for events in a non-blocking fashion.
+
+I didn't want to do significant changes to either libusb or gPhoto2, and I've already used Asyncify for `Promise` integration earlier, so that's the path I've chosen. In order to simulate a blocking variant of `poll`, for the initial proof of concept I've used a loop below:
+
+```cpp
+#ifdef __EMSCRIPTEN__
+  // TODO: optimize this. Right now it will keep unwinding-rewinding the stack
+  // on each short sleep until an event comes or the timeout expires.
+  // We should probably create an actual separate thread that does signaling
+  // or come up with a custom event mechanism to report events from
+  // `usbi_signal_event` and process them here.
+  double until_time = emscripten_get_now() + timeout_ms;
+  do {
+    // Emscripten `poll` ignores timeout param, but pass 0 explicitly just
+    // in case.
+    num_ready = poll(fds, nfds, 0);
+    if (num_ready != 0) break;
+    // Yield to the browser event loop to handle events.
+    emscripten_sleep(0);
+  } while (emscripten_get_now() < until_time);
+#else
+  num_ready = poll(fds, nfds, timeout_ms);
+#endif
+```
+
+What it does is:
```

Comment:
Is this actually a sequence? Many use numbered bullets for things that aren't. If this list isn't a sequence it should be bullets instead of numbers.

Reply:
It is.